File: pnl_analyzer/protocols.py
from pydantic import BaseModel, ConfigDict, AliasGenerator
from pydantic.alias_generators import to_camel


import numpy as np

class Protocol(BaseModel):
    model_config = ConfigDict(
        alias_generator=AliasGenerator(serialization_alias=to_camel)
    )


class PlotData(Protocol):
    x: list
    y: list
    type: str = "scatter"  # Default to scatter plot
    mode: str = "lines"
    marker: dict = {"color": "red"}
    name: str | None = None


class PlotLayout(Protocol):
    title: dict = {"text": "My Plot"}
    xaxis: dict = {"title": "X-Axis"}
    yaxis: dict = {"title": "Y-Axis"}
    autosize: bool = True


class PlotResponse(BaseModel):
    data: list[PlotData] = []# A list of traces
    layout: PlotLayout = PlotLayout()

File: webapp/src/protocols.ts
export interface PlotData {
  x: any[];
  y: any[];
  type?: string;
  mode?: string;
  marker?: any;
  name?: string;
  stackgroup?: string;
  fill?: string;
}


export interface PlotLayout {
  title: string;
  xaxis?: any;
  yaxis?: any;
  autosize: boolean;
}


export interface PlotResponse {
  data: PlotData[];
  layout: PlotLayout;
}
